File: findCircleNum.java
import java.util.*;
// Approach 1- DFS 
class Solution{ // TC O(n^2) SC O(n)
    public void dfs(int node, int[][] isConnected, boolean[] visit) {
        visit[node] = true;
        for (int i = 0; i < isConnected.length; i++) {
            if (isConnected[node][i] == 1 && !visit[i]) {
                dfs(i, isConnected, visit);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        int numberOfComponents = 0;
        boolean[] visit = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!visit[i]) {
                numberOfComponents++;
                dfs(i, isConnected, visit);
            }
        }

        return numberOfComponents;
    }
}

// Approach 2- BFS 
class Solution1{ // TC O(n^2) SC O(n)
    public void bfs(int node, int[][] isConnected, boolean[] visit) {
        Queue<Integer> q = new LinkedList<>();
        q.offer(node);
        visit[node] = true;

        while (!q.isEmpty()) {
            node = q.poll();

            for (int i = 0; i < isConnected.length; i++) {
                if (isConnected[node][i] == 1 && !visit[i]) {
                    q.offer(i);
                    visit[i] = true;
                }
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;
        int numberOfComponents = 0;
        boolean[] visit = new boolean[n];

        for (int i = 0; i < n; i++) {
            if (!visit[i]) {
                numberOfComponents++;
                bfs(i, isConnected, visit);
            }
        }

        return numberOfComponents;
    }
}

// Approach 3- Union-Find 
class Solution2{ // TC O(n^2) SC O(n)
    public int findCircleNum(int[][] isConnected) {
        int n=isConnected.length;
        UnionFind uf=new UnionFind(n);
        int answer=n;
        
        for(int i=0; i<n; i++){
            for(int j=0; j<n; j++){
                if(isConnected[i][j]==1 && uf.find(i)!=uf.find(j)){
                    answer--;
                    uf.union(i,j);
                }
            }
        }
        return answer;
    }
}

class findCircleNum{
    public static void main(String[] args) {
        int[][] isConnected={{1,1,0},{1,1,0},{0,0,1}};
        Solution s= new Solution();
        System.out.println(s.findCircleNum(isConnected));
        Solution1 s1= new Solution1();
        System.out.println(s1.findCircleNum(isConnected));
        Solution2 s2= new Solution2();
        System.out.println(s2.findCircleNum(isConnected));
    }
}

File: findCircleNum.py
from implementUnionFind import *  
from collections import deque
class Solution:
# Approach 1- DFS
	def findCircleNum(self, isConnected): # TC O(n^2) // SC O(n)
	    n=len(isConnected)
	    visited=[False]*n
	    
	    def dfs(node):
	        visited[node]=True
	        for i in range(n):
	            if isConnected[node][i]==1 and not visited[i]:
	                dfs(i)
	    
	    answer=0
	    for i in range(n):
	        if not visited[i]:
	            dfs(i)
	            answer+=1
	    
	    return answer

# Approach 2- BFS
	def findCircleNum1(self, isConnected): # TC O(n^2) // SC O(n)
		n=len(isConnected)
		visited=[False]*n

		def bfs(node):
		    queue=deque()
		    queue.append(node)
		    visited[node]=True
		    while queue:
		        node=queue.popleft()
		        
		        for i in range(n):
		            if isConnected[node][i]==1 and not visited[i]:
		                queue.append(i)
		                visited[i]=True

		answer=0
		for i in range(n):
		    if not visited[i]:
		        bfs(i)
		        answer+=1

		return answer

# Approach 3- Union-Find
	def findCircleNum2(self, isConnected): # TC O(n^2) // SC O(n)
		n=len(isConnected)
		uf=UnionFind(n)
		answer=n

		for i in range(n):
			for j in range(n):
				if isConnected[i][j]==1 and uf.find(i)!=uf.find(j):
					answer-=1
					uf.union(i,j)

		return answer

s=Solution()
print(s.findCircleNum([[1,0,0],[0,1,0],[0,0,1]]))
print(s.findCircleNum1([[1,0,0],[0,1,0],[0,0,1]]))
print(s.findCircleNum2([[1,0,0],[0,1,0],[0,0,1]]))
